doc/rtd/topics/boot.rst

```diff
@@ -157,4 +157,93 @@ finished, the ``cloud-init status`` subcommand can help block external
 units dependency chains. See :ref:`cli_status` for more info.
 
+First Boot Determination
+************************
+
+cloud-init has to determine whether or not the current boot is a first boot or
+not, so that it applies the appropriate configuration.  On a first boot, it
+should run all "per-instance" configuration, whereas on a subsequent boot it
+should run only "per-boot" configuration.  This section describes how
+cloud-init performs this determination, as well as why it is necessary.
+
+When it runs, cloud-init stores a cache of its internal state for use across
+stages and boots.
+
+If this cache is present, then cloud-init has run on this system before.
+[#not-present]_  There are two cases where this could occur.  Most commonly,
+the instance has been rebooted, and this is a second/subsequent boot.
+Alternatively, the filesystem has been attached to a *new* instance, and this
+is a first boot.  The most obvious case where this happens is when an instance
+is launched from an image captured from a launched instance.
+
+By default, cloud-init attempts to determine which case we're running in by
+checking the instance ID in the cache against the instance ID it determines at
+runtime.  If they do not match, then this is a first boot; otherwise, it's a
+subsequent boot.  Internally, cloud-init refers to this behavior as ``check``.
+
+This behavior is required for images captured from launched instances to
+behave correctly, and so is the default which generic cloud images ship with.
+However, there are cases where it can cause problems. [#problems]_ For these
+cases, cloud-init has support for modifying its behavior to trust the instance
+ID that is present in the system unconditionally.  This means that cloud-init
+will never detect a first boot when the cache is present, and it follows that
+the only way to cause cloud-init to detect a first boot is to manually remove
+cloud-init's cache.  Internally, this behavior is referred to as ``trust``.
+
+To configure which of these behaviors to use, cloud-init exposes the
+``manual_cache_clean`` configuration option.  When ``false``, cloud-init will
+``check`` and clean the cache if the instance IDs do not match (this is the
+default, as discussed above).  When ``true``, cloud-init will ``trust`` the
+existing cache (and therefore not clean it).
+
+Manual Cache Cleaning
+=====================
+
+cloud-init looks in ``/var/lib/cloud/instance/obj.pkl`` for its cache.
+``/var/lib/cloud/instance`` is a symlink to a subdirectory of
+``/var/lib/cloud/instances/``, which contains a single directory per instance
+that has booted with this filesystem.  To clean the cache in the least
+destructive manner possible, simply remove the ``/var/lib/cloud/instance``
+symlink.
+
+.. note:: If you do want to destroy all cached data, then in addition to
+   removing the ``/var/lib/cloud/instance`` symlink, you should also remove
+   all directories under ``/var/lib/cloud/instances``.
```

Comment:
I don't really like documenting this "interface".

for "all cached data", this should just be `cloud-init clean`.

I'm not sure if we need to mention to a user the ability to just remove the symlink or not.  Can you think of any non-developer use where that is valuable?

Reply:
Agreed, I've just changed it to reference clean.